### ISSUE TITLE
Multi-fields will create field names with dots '.'

### ISSUE THREAD

**Issue by brusic** [elastic/elasticsearch#14957]
One of the breaking changes of Elasticsearch 2.0 was that field names are no longer allowed to have dots: https://github.com/elastic/elasticsearch/pull/12068

However, if you use the new multi-field syntax, Elasticsearch will create field name with dots. The documentation supports the behavior. The aggregation in the example is on a field named "city.raw":

https://www.elastic.co/guide/en/elasticsearch/reference/current/multi-fields.html

The unit tests in the code look for field names with dots: https://github.com/elastic/elasticsearch/blob/2.1/core/src/test/java/org/elasticsearch/index/mapper/multifield/MultiFieldTests.java#L88

The tests pass because the check in ObjectMapper$TypeParser only checks the original name.

The field name is built in ContentPath:
https://github.com/elastic/elasticsearch/blob/2.1/core/src/main/java/org/elasticsearch/index/mapper/ContentPath.java#L50
https://github.com/elastic/elasticsearch/blob/2.1/core/src/main/java/org/elasticsearch/index/mapper/ContentPath.java#L80-L87

These field names will fail if the indices are run through the MetaDataIndexUpgradeService. Are dots allowed in this context?

**Comment by clintongormley** [comment]
well spotted, thanks!  Simple recreation, that runs on 2.1 (or on 1.7 and upgrades without warning to 2.1):

    PUT t
    {
      "mappings": {
        "t": {
          "properties": {
            "foo": {
              "type": "string",
              "fields": {
                "bar.baz": {
                  "type": "string",
                  "index": "not_analyzed"
                }
              }
            }
          }
        }
      }
    }
    
    PUT t/t/1
    {
      "foo": "test"
    }
    
    GET t/t/_search
    {
      "query": {
        "match": {
          "foo.bar.baz": "test"
        }
      }
    }

**Comment by dadoonet** [comment]
I think it's the same root cause as described for the mapper attachment plugin. Linking to it here https://github.com/elastic/elasticsearch-mapper-attachments/issues/169

**Comment by brwe** [comment]
I took a look and must admit I do not fully understand what the issue is.

@brusic What exactly is the error you see? I wrote a test for the example and all seems well (https://github.com/brwe/elasticsearch/commit/43e94268e4d59d85b46fb204275d3b24c288bad6#diff-523d9598b02bfef9341f7aa20ebfdf47R543). The ContentPath builds the field path not the name and if I understand correctly that contains as many dots as there are levels.
I think I am missing something?

@clintongormley We should not be able to create multi fields with dots in the name, that is a bug for sure. I'll try to fix

**Comment by brusic** [comment]
It all depends on how strict should Elasticsearch should be regarding non-dots-in-field names. Better yet, what is the difference between a field name and a path?

At the Lucene level, multifield will create a field name with a dot in it. Inspected a one-shard index with Luke, and the fields created using the fields syntax will have a dot. Clinton's example does not fully show the impact since it uses a field name with a dot. Using 'bar' instead of 'bar.baz'.

{
  "mappings": {
    "t": {
      "properties": {
        "foo": {
          "type": "string",
          "fields": {
            "bar": {
              "type": "string",
              "index": "not_analyzed"
            }
          }
        }
      }
    }
  }
}

PUT t/t/1
{
  "foo": "test"
}

GET t/t/_search
{
  "query": {
    "match": {
      "foo.bar": "test"
    }
  }
}

A field name of foo.bar was created. Is this field name allowed? There was a discussion on the mailing list were these field names were getting flagged as incorrect when a user attempted to upgrade:

https://discuss.elastic.co/t/elasticsearch-2-mapping/34366/9

Either dots should never be allowed or the fields created with multifields should not be flagged as incorrect since they are paths and not field names.

**Comment by rjernst** [comment]
> A field name of foo.bar was created

This isn't true. We do not create a field with name `foo.bar`. We create the equivalent of the following mapping:

```
{
  "mappings": {
    "t": {
      "properties": {
        "foo": {
          "type": "string",
          "copy_to": "foo.bar",
          "properties": {
            "bar": {
              "type": "string",
              "index": "not_analyzed"
            }
          }
        }
      }
    }
  }
}
```

The `name` of a field is *just* the name. In this case, the `name` of the multi field is `bar`.  It's path is how you access it in a query, which uses dots to separate path elements. So `bar` is underneath a `foo` field, and so its path is `foo.bar`. This path is why dots are no longer allowed in field names: we cannot distinguish between `foo.bar` being a field on its own, or an object field `foo` with a subfield `bar`.

**Comment by rjernst** [comment]
FYI, the mapping I showed there wont' actually work IIRC. It was just used to explain how multi fields work conceptually.  Only an `object` field can have `properties` (ie subfields) hence why the general purpose `fields` exists for concrete data types.

**Comment by brwe** [comment]
I think the confusion stems from the ambiguous use of the words `field name`. We have 1) a _Lucene_ field name which can contain as many dots as needed (we create them for example [here](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/index/mapper/core/StringFieldMapper.java#L306) and this is also what we can see in Luke) and 2) a `field name` in _elasticsearch mapping_ which is just the innermost identifier for a value in a nested json structure and may not contain any dots in elasticsearch mapping.

For example:

```
{
  "a": {
     "b": {
        "c":  "All mappings and no play makes Britta a dull programmer"
     }
  }
}
```
1) will be `a.b.c`
2) will be `c` 

The path we were talking about before is `a.b` and is needed on Lucene level I think to distinguish `a.b.c` for example from another field that also has the name `c` but is nested in objects `d` and `e` (and therefore has path `d.e`). 
Let me know if this makes sense.

Also, @brusic just to be sure: did you actually encounter any error? We only check the elasticsearch field name for dots, the Lucene field names are untouched by that.

**Comment by rjernst** [comment]
>   The path we were talking about before is a.b

One minore correction: path would be `a.b.c` there. It is like path vs name for File in java.

**Comment by brusic** [comment]
I agree with Britta in that the confusion is related to the ambiguity of the words field name, which I acknowledged in my previous comment. I have not seen the error myself, but have seen it a couple of times on the forum. Since I was curious what was happening behind the scenes, I checked it out myself, only looking at the Lucene field names.

If someone else has a problem with upgrading, they can create a new issue.
